File: ts/src/cmd/push.ts
import { RemoteClient } from '../remote';
import { RadStore } from '../store';
import * as fs from 'fs';
import * as path from 'path';

interface RemoteConfig {
  url: string;
}

export async function runPush(store: RadStore): Promise<string> {
  // Load remote.json
  const remotePath = path.join('.rad', 'remote.json');
  const remoteJson = fs.readFileSync(remotePath, 'utf-8');
  const remoteConfig: RemoteConfig = JSON.parse(remoteJson);

  const client = new RemoteClient(remoteConfig.url);

  // Load operations
  const operations = store.loadOplog();

  // Track pushed and skipped operations
  const pushedOps: Array<{ id: string; status: string; regionId: string }> = [];
  let skipped = 0;

  // Submit each operation
  for (const op of operations) {
    try {
      await client.submitOperation(op);
      pushedOps.push({ id: op.id, status: op.status, regionId: op.regionId });
    } catch (error) {
      const errorMsg = error instanceof Error ? error.message : String(error);
      // If it's a duplicate, that's okay (idempotency)
      if (errorMsg.includes('409') || errorMsg.includes('already exists')) {
        skipped++;
      } else {
        // Other errors we should report but not fail
        console.error(`Warning: Failed to push ${op.id}: ${errorMsg}`);
      }
    }
  }

  // Build output
  let output = '';
  output += `pushed: ${pushedOps.length} operations to ${remoteConfig.url}\n`;

  for (const { id, status, regionId } of pushedOps) {
    // Extract file from regionId
    const colonPos = regionId.indexOf(':');
    const file = colonPos !== -1 ? regionId.substring(0, colonPos) : regionId;

    output += `  ${id} [${status}] ${file}\n`;
  }

  if (skipped > 0) {
    output += `  (${skipped} operations already on server)\n`;
  }

  return output;
}

File: rust/src/cmd/push.rs
use crate::remote::RemoteClient;
use crate::store::RadStore;
use std::fs;

#[derive(serde::Deserialize)]
struct RemoteConfig {
    url: String,
}

pub fn run_push(store: &RadStore) -> Result<String, String> {
    // Load remote.json
    let remote_path = std::env::current_dir()
        .map_err(|e| format!("Failed to get current directory: {}", e))?
        .join(".rad/remote.json");

    let remote_json = fs::read_to_string(&remote_path)
        .map_err(|e| format!("Failed to read remote.json: {}", e))?;

    let remote_config: RemoteConfig = serde_json::from_str(&remote_json)
        .map_err(|e| format!("Failed to parse remote.json: {}", e))?;

    let client = RemoteClient::new(&remote_config.url);

    // Load operations
    let oplog = store.load_oplog()?;
    let operations = oplog.get_all_operations();

    // Track pushed and skipped operations
    let mut pushed_ops = Vec::new();
    let mut skipped = 0;

    // Submit each operation
    for op in operations {
        // Skip if already accepted (these were likely from clone/pull)
        // We only push operations that are local and not yet on the server
        // For simplicity, we'll try to push all operations and let the server handle duplicates

        match client.submit_operation(&op) {
            Ok(_) => {
                pushed_ops.push((op.id.clone(), op.status.clone(), op.region_id.clone()));
            }
            Err(e) => {
                // If it's a duplicate, that's okay (idempotency)
                if e.contains("409") || e.contains("already exists") {
                    skipped += 1;
                } else {
                    // Other errors we should report but not fail
                    eprintln!("Warning: Failed to push {}: {}", op.id, e);
                }
            }
        }
    }

    // Build output
    let mut output = String::new();
    output.push_str(&format!("pushed: {} operations to {}\n", pushed_ops.len(), remote_config.url));

    for (id, status, region_id) in &pushed_ops {
        // Extract file from region_id
        let file = if let Some(colon_pos) = region_id.find(':') {
            &region_id[..colon_pos]
        } else {
            region_id.as_str()
        };

        output.push_str(&format!("  {} [{:?}] {}\n", id, status, file));
    }

    if skipped > 0 {
        output.push_str(&format!("  ({} operations already on server)\n", skipped));
    }

    Ok(output)
}
